File: problems/968/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

enum Camera {
    Covered,
    HasCamera,
    NeedsCover
};

class Solution {
public:
    int minCameraCover(TreeNode* root) {
        int ans = 0;
        
        if (helper(root, ans) == NeedsCover) {
            ans++;
        }
        
        return ans;
    }
    
    Camera helper(TreeNode *root, int& ans) {
        if (not root) {
            return Covered;
        }
        
        Camera left = helper(root->left, ans),
            right = helper(root->right, ans);
        
        if (left == NeedsCover or right == NeedsCover) {
            ans++;
            return HasCamera;
        }
        
        if (left == HasCamera or right == HasCamera) {
            return Covered;
        }
        
        return NeedsCover;
    }
};

File: problems/968/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Camera int

const (
    HasCam Camera = iota
    Covered
    NotCovered
)

func minCameraCover(root *TreeNode) int {
    ans := 0
    
    if helper(root, &ans) == NotCovered {
        return ans + 1
    }
    
    return ans
}

func helper(root *TreeNode, ans *int) Camera {
    if root == nil {
        return Covered
    }
    
    l, r := helper(root.Left, ans), helper(root.Right, ans)
    
    if l == NotCovered || r == NotCovered {
        (*ans)++
        return HasCam
    }
    
    if l == HasCam || r == HasCam {
        return Covered
    }
    
    return NotCovered
}